src/main/kotlin/ru/hse/spb/Visitor.kt

```diff
@@ -0,0 +1,250 @@
+package ru.hse.spb
+
+import ru.hse.spb.parser.FunCallBaseVisitor
+import ru.hse.spb.parser.FunCallParser
+import java.lang.IllegalArgumentException
+import kotlin.reflect.KFunction
+import kotlin.reflect.full.declaredMemberFunctions
+import kotlin.reflect.full.declaredMemberProperties
+
+class Space(val parent: Space? = null) {
+    private val functions = mutableMapOf<String, FunCallParser.FunctionContext>()
+    private val variables = mutableMapOf<String, Int?>()
+
+    fun createChild(): Space {
+        return Space(this)
+    }
+
+    fun findFunction(name: String): FunCallParser.FunctionContext? {
+        return functions[name] ?: parent?.findFunction(name)
+    }
+
+    fun rewriteVariable(name: String, value: Int) {
+        if (variables.containsKey(name)) {
+            variables[name] = value
+            return
+        }
+        parent?.rewriteVariable(name, value) ?: throw IllegalArgumentException()
+    }
+
+    fun findVariable(name: String): Int? {
+        return variables[name] ?: parent?.findVariable(name)
+    }
+
+    fun addFunction(func: FunCallParser.FunctionContext) {
+        val name = func.IDENTIFIER().text
+        require(findFunction(name) == null)
+        functions[name] = func
+    }
+
+    fun addVariable(name: String, value: Int?) {
+        require(!variables.containsKey(name))
+        variables[name] = value
+    }
+
+    fun extend(func: FunCallParser.FunctionContext, arguments: List<Int>): Space {
+        val argumentNames = func.argumentList().toNames()
+        val space = createChild()
+        for ((name, exp) in argumentNames.zip(arguments)) {
+            space.variables[name] = exp
+        }
+        return space
+    }
+}
+
+class Visitor : FunCallBaseVisitor<Unit>() {
+    var namespace = Space()
+    val stack = mutableListOf<Int>()
+    var exit = false
+
+    override fun visitFile(ctx: FunCallParser.FileContext) {
+        ctx.block().accept(this)
+    }
+
+    override fun visitBlock(ctx: FunCallParser.BlockContext) {
+        for (statement in ctx.statement()) {
+            statement.accept(this)
+            if (exit) {
+                exit = false
+                return
+            }
+        }
+        if (stack.isEmpty()) {
+            stack.add(0)
+        }
+    }
+
+    override fun visitStatement(ctx: FunCallParser.StatementContext) {
+        val list = listOf(ctx.functionCall(),
+                ctx.assigment(),
+                ctx.derji(),
+                ctx.esli(),
+                ctx.poka(),
```

Comment:
[:are-you-kidding-me:](https://www.google.com/search?q=are+you+kidding+me&rlz=1C5CHFA_enRU739RU740&source=lnms&tbm=isch&sa=X&ved=2ahUKEwjOh9-n5I_mAhVPxaYKHdK0BsoQ_AUoAXoECA4QAw&biw=1440&bih=766#imgrc=6sGKLRtfRmk82M:) :)